### ISSUE TITLE
Undefined references in KokkosKernels nightly builds when both Pthread and Serial exec spaces enabled

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#4625]
Following changes for deprecation of `Kokkos_ENABLE_PTHREAD` #4619 KokkosKernels builds with both Pthread and Serial backends enabled began failing due to undefined references; this does not happen when enabling only the Pthread or Serial backend, and does not show up in Kokkos testing with Pthread and Serial backends enabled.

**Reproducer**:
```
module load sems-archive-env sems-archive-cmake/3.17.1 sems-archive-gcc/9.2.0

$KOKKOSKERNELS_PATH/cm_generate_makefile.bash --with-devices=Pthread,Serial --arch=SNB,Volta70 --compiler=g++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized " --cxxstandard="14" --kokkos-path=$KOKKOS_PATH --kokkoskernels-path=$KOKKOSKERNELS_PATH --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft --with-options=disable_deprecated_code
```

**Comment by ndellingwood** [comment]
Here are the corresponding CMake lines from the generate_makefile script:

```
# Kokkos
cmake -DCMAKE_CXX_COMPILER=g++ -DCMAKE_CXX_FLAGS="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized"  -DCMAKE_INSTALL_PREFIX=/ascldap/users/ndellin/kokkos-kernels/testing/Gcc91/kokkos-install -DKokkos_ENABLE_SERIAL=ON -DKokkos_ENABLE_PTHREAD=ON -DKokkos_ARCH_VOLTA70=ON -DKokkos_ARCH_SNB=ON -DKokkos_ENABLE_TESTS=OFF -DKokkos_ENABLE_EXAMPLES=OFF -DKokkos_ENABLE_DEPRECATED_CODE=OFF -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF -DCMAKE_CXX_STANDARD=14 -DKokkos_ENABLE_DEPRECATION_WARNINGS=OFF $KOKKOS_PATH

# Install Kokkos

# KokkosKernels
make -DCMAKE_CXX_COMPILER=g++ -DKokkos_DIR=/ascldap/users/ndellin/kokkos-kernels/testing/Gcc91/kokkos-install/lib64/cmake/Kokkos -DCMAKE_CXX_FLAGS="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized" -DKokkosKernels_ENABLE_TESTS_AND_PERFSUITE=OFF -DKokkosKernels_ENABLE_TESTS=ON -DKokkosKernels_ENABLE_EXAMPLES:BOOL=ON -DKokkosKernels_INST_COMPLEX_DOUBLE=ON -DKokkosKernels_INST_DOUBLE=ON -DKokkosKernels_INST_ORDINAL_INT=ON -DKokkosKernels_INST_OFFSET_SIZE_T=ON -DKokkosKernels_INST_OFFSET_INT=ON -DKokkosKernels_INST_LAYOUTLEFT=ON -DKokkosKernels_ENABLE_TPL_CUSPARSE=OFF -DKokkosKernels_ENABLE_TPL_CUBLAS=OFF $KOKKOSKERNELS_PATH

# Build KokkosKernels, undefined refs when linking unit tests
```

**Comment by dalg24** [comment]
Don't you get warning about `Kokkos_ENABLE_DEPRECATED_CODE` not being used.  I am pretty sure we only recognize `Kokkos_ENABLE_DEPRECATED_CODE_3`

**Comment by dalg24** [comment]
Assuming that deprecated code is OFF, we do not export `Kokkos_ENABLE_PTRHEAD` any more so that is probably where to start looking.

**Comment by ndellingwood** [comment]
I rebuilt KokkosKernels with changes in the build system that seem to have resolved this:

```
[ndellin@kokkos-dev-2 Gcc91]$ git diff
diff --git a/cmake/KokkosKernelsConfig.cmake.in b/cmake/KokkosKernelsConfig.cmake.in
index f930dd5..9bdce0e 100644
--- a/cmake/KokkosKernelsConfig.cmake.in
+++ b/cmake/KokkosKernelsConfig.cmake.in
@@ -14,7 +14,7 @@ SET(Kokkos_ENABLE_OPENMPTARGET @Kokkos_ENABLE_OPENMPTARGET@)
 SET(Kokkos_ENABLE_CUDA         @Kokkos_ENABLE_CUDA@)
 SET(Kokkos_ENABLE_HIP          @Kokkos_ENABLE_HIP@)
 SET(Kokkos_ENABLE_SYCL         @Kokkos_ENABLE_SYCL@)
-SET(Kokkos_ENABLE_PTHREAD      @Kokkos_ENABLE_PTHREAD@) 
+SET(Kokkos_ENABLE_THREADS      @Kokkos_ENABLE_THREADS@) 
 SET(Kokkos_ENABLE_SERIAL       @Kokkos_ENABLE_SERIAL@)
 
 INCLUDE("${KokkosKernels_CMAKE_DIR}/KokkosKernelsTargets.cmake")
diff --git a/cmake/kokkos_backends.cmake b/cmake/kokkos_backends.cmake
index 11c51ee..a90ad69 100644
--- a/cmake/kokkos_backends.cmake
+++ b/cmake/kokkos_backends.cmake
@@ -10,7 +10,7 @@ MACRO(CHECK_KOKKOS_BACKEND BE)
 ENDMACRO(CHECK_KOKKOS_BACKEND)
 
 CHECK_KOKKOS_BACKEND(SERIAL)
-CHECK_KOKKOS_BACKEND(PTHREAD)
+CHECK_KOKKOS_BACKEND(THREADS)
 CHECK_KOKKOS_BACKEND(OPENMP)
 CHECK_KOKKOS_BACKEND(OPENMPTARGET)
 CHECK_KOKKOS_BACKEND(CUDA)
diff --git a/cmake/kokkoskernels_eti_devices.cmake b/cmake/kokkoskernels_eti_devices.cmake
index 81ab895..47dce1f 100644
--- a/cmake/kokkoskernels_eti_devices.cmake
+++ b/cmake/kokkoskernels_eti_devices.cmake
@@ -8,7 +8,7 @@ SET(EXEC_SPACES
   EXECSPACE_SYCL
   EXECSPACE_OPENMPTARGET
   EXECSPACE_OPENMP
-  EXECSPACE_PTHREAD
+  EXECSPACE_THREADS
   EXECSPACE_SERIAL
 )
 SET(EXECSPACE_CUDA_CPP_TYPE         Kokkos::Cuda)
@@ -16,7 +16,7 @@ SET(EXECSPACE_HIP_CPP_TYPE          Kokkos::Experimental::HIP)
 SET(EXECSPACE_SYCL_CPP_TYPE         Kokkos::Experimental::SYCL)
 SET(EXECSPACE_OPENMPTARGET_CPP_TYPE Kokkos::Experimental::OpenMPTarget)
 SET(EXECSPACE_OPENMP_CPP_TYPE       Kokkos::OpenMP)
-SET(EXECSPACE_PTHREAD_CPP_TYPE      Kokkos::Threads)
+SET(EXECSPACE_THREADS_CPP_TYPE      Kokkos::Threads)
 SET(EXECSPACE_SERIAL_CPP_TYPE       Kokkos::Serial)
 
 SET(MEM_SPACES
@@ -174,17 +174,17 @@ ENDIF()
 
 KOKKOSKERNELS_ADD_OPTION(
   INST_EXECSPACE_THREADS
-  ${KOKKOSKERNELS_INST_EXECSPACE_PTHREAD_DEFAULT}
+  ${KOKKOSKERNELS_INST_EXECSPACE_THREADS_DEFAULT}
   BOOL
-  "Whether to build kernels for the execution space Kokkos::Threads.  If explicit template instantiation (ETI) is enabled in Trilinos, disabling this when Kokkos_ENABLE_PTHREAD is ena
+  "Whether to build kernels for the execution space Kokkos::Threads.  If explicit template instantiation (ETI) is enabled in Trilinos, disabling this when Kokkos_ENABLE_THREADS is ena
 )
 #There continues to be name ambiguity with threads vs pthreads
-SET(KOKKOSKERNELS_INST_EXECSPACE_PTHREAD ${KOKKOSKERNELS_INST_EXECSPACE_THREADS})
+SET(KOKKOSKERNELS_INST_EXECSPACE_THREADS ${KOKKOSKERNELS_INST_EXECSPACE_THREADS})
 
-IF(KOKKOSKERNELS_INST_EXECSPACE_PTHREAD AND KOKKOSKERNELS_INST_MEMSPACE_HOSTSPACE)
+IF(KOKKOSKERNELS_INST_EXECSPACE_THREADS AND KOKKOSKERNELS_INST_MEMSPACE_HOSTSPACE)
   LIST(APPEND DEVICE_LIST "<Threads,HostSpace>")
-  IF(NOT KOKKOS_ENABLE_PTHREAD)
-    MESSAGE(FATAL_ERROR "Set ETI on for PTHREAD, but Kokkos was not configured with the PTHREAD backend")
+  IF(NOT KOKKOS_ENABLE_THREADS)
+    MESSAGE(FATAL_ERROR "Set ETI on for THREADS, but Kokkos was not configured with the THREADS backend")
   ENDIF()
 ENDIF()
 
@@ -201,7 +201,7 @@ SET(EXECSPACE_SYCL_VALID_MEM_SPACES               SYCLSPACE SYCLSHAREDSPACE)
 SET(EXECSPACE_OPENMPTARGET_VALID_MEM_SPACES       OPENMPTARGETSPACE)
 SET(EXECSPACE_SERIAL_VALID_MEM_SPACES             HBWSPACE HOSTSPACE)
 SET(EXECSPACE_OPENMP_VALID_MEM_SPACES             HBWSPACE HOSTSPACE)
-SET(EXECSPACE_PTHREAD_VALID_MEM_SPACES            HBWSPACE HOSTSPACE)
+SET(EXECSPACE_THREADS_VALID_MEM_SPACES            HBWSPACE HOSTSPACE)
 SET(DEVICES)
 FOREACH(EXEC ${EXEC_SPACES})
   IF (KOKKOSKERNELS_INST_${EXEC})
diff --git a/unit_test/CMakeLists.txt b/unit_test/CMakeLists.txt
index f226063..59498ef 100644
--- a/unit_test/CMakeLists.txt
+++ b/unit_test/CMakeLists.txt
@@ -367,7 +367,7 @@ IF (KOKKOS_ENABLE_SERIAL)
   )
 ENDIF ()
 
-IF (KOKKOS_ENABLE_PTHREAD)
+IF (KOKKOS_ENABLE_THREADS)
   KOKKOSKERNELS_INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR}/threads)
   KOKKOSKERNELS_INCLUDE_DIRECTORIES(REQUIRED_DURING_INSTALLATION_TESTING ${CMAKE_CURRENT_SOURCE_DIR}/threads)
```

I'll put in a PR to kokkos-kernels with the updates and can close this issue.
@dalg24 kind of strange that this only occurs when both Serial and Pthreads are simultaneously enabled, but not individually, right?

**Comment by ndellingwood** [comment]
Cross-reference kokkos/kokkos-kernels#1237

**Comment by ndellingwood** [comment]
This was resolved with kokkos/kokkos-kernels#1249
